ports/nrf/modules/machine/uart.c

```diff
@@ -355,9 +379,20 @@ STATIC mp_uint_t machine_hard_uart_write(mp_obj_t self_in, const void *buf_in, m
 STATIC mp_uint_t machine_hard_uart_ioctl(mp_obj_t self_in, mp_uint_t request, uintptr_t arg, int *errcode) {
     machine_hard_uart_obj_t *self = self_in;
     (void)self;
+    mp_uint_t ret = 0;
 
-    if (request == MP_STREAM_FLUSH) {
-        // Since uart.write() waits up to the last byte, uart.flush() always succeds.
+    if (request == MP_STREAM_POLL) {
+        uintptr_t flags = arg;
+        if ((flags & MP_STREAM_POLL_RD) && uart_rx_any(self) != 0) {
+            ret |= MP_STREAM_POLL_RD;
+        }
+        if ((flags & MP_STREAM_POLL_WR) && machine_uart_txdone(self) == mp_const_true) {
```

Comment:
as above, probably best to make a helper `machine_uart_tx_in_progress()` (can also be used just below)